plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp

```diff
@@ -205,12 +205,12 @@ class bp_connection_manager {
       return conn->bp_connection == Connection::bp_connection_type::non_bp && conn->socket_is_open() && conn->incoming_and_handshake_received();
    }
 
-   send_buffer_type get_gossip_bp_initial_send_buffer() {
+   const send_buffer_type& get_gossip_bp_initial_send_buffer() {
```

Comment:
For thread-safety this needs to remain `send_buffer_type`.

Reply:
For my enlightenment, please explain why?

Reply:
`bp_connection_manager`

thread A:
calls `get_gossip_bp_initial_send_buffer()`
acquires mutex mtx
assigns reference `send_buffer` to caller
releases mutex

thread B:
calls `update_bp_producer_peers()`
acquires mutex mtx
calls `initial_gossip_msg_factory.set_initial_send_buffer(signed_empty);`
starts to update `send_buffer`

thread A:
starts to use `send_buffer`

Reply:
Thanks. Missed that.
Still not 100% correct to read a value in one thread (`send_buffer`) which can be modified in another thread. 
maybe `send_buffer` should be an `atomic<shared_ptr>` in `gossip_buffer_initial_factory`

Reply:
> Still not 100% correct to read a value in one thread (`send_buffer`) which can be modified in another thread. maybe `send_buffer` should be an `atomic<shared_ptr>` in `gossip_buffer_initial_factory`

Oh wait I see we hold the mtx, so we are fine indeed.

Reply:
That shouldn't be possible. A copy of the shared_ptr is made before releasing the mutex. The original shared_ptr is updated with new data, but that doesn't affect the copy.

`std::shared_ptr` - "All member functions (including copy constructor and copy assignment) can be called by multiple threads on different shared_ptr objects without additional synchronization even if these objects are copies and share ownership of the same object."